Update Armagetron to 0.2.9.1.0 and runtime to 20.08, continued

zmanuel opened the pull request:
Hi, upstream developer here, dealing with my own mess. Taking over from the fine work in #4.

On top of the updating work, the big change is that I'm using the upsteam package's ability to change its name completely to org.armagetronad.ArmagetronAdvanced via the progname variable; we typically use that to get our testing builds installable in parallel with the stable version and to rename the game for trademark sensitive environments (we're ArmaCycles on Red Hat and RetroCycles on Steam, because Arma is also taken...). Anyway, doing this should put everything in the right places with the right name and content.

Except it doesn't, because we messed up and end up with things named org.armagetronad.org.armagetronad.ArmagetronAdvanced. Something to correct later.

And while the icons are installed correctly, flatpak only seems to picks them up for export if you use the 'install' command, which we don't. So more mending is required.

AsciiWolf committed:
Update Armagetron to 0.2.9.1.0 and runtime to 20.08

zmanuel committed:
Finish update to 0.2.9.1

Let game itself rename to org.armagetronad.ArmagetronAdvanced,
prevents headaches and later adaptions

Add standard user configuration and storage directory to sandbox
permissions

Disable generating superfluous uninstall script

Restore installation adaptions that are still required;
Icons need an explicit install to be picked up by the exporter,
new kerfuffle with appstream metadata needs compensating

AsciiWolf commented on org.armagetronad.ArmagetronAdvanced.json:
```diff
@@ -8,10 +8,12 @@
         "--share=network",
-        "--socket=pulseaudio"
+        "--socket=pulseaudio",
+        "--filesystem=~/.org.armagetronad.ArmagetronAdvanced/var:create",
```
This is quite nonstandard directory. I wonder what is wrong with the current one inside `~/.var/app/org.armagetronad.ArmagetronAdvanced`? :-)

zmanuel replied:
That it didn't work properly, see #3. But yes, we have command line arguments to change those directories to a more modern place and should use them here. ~/.appname is so 2001.

AsciiWolf replied:
Ah, I just found #3 as well. :-) Anyway, `--persist=.org.armagetronad.ArmagetronAdvanced` (note that there is no `~/` prefix used) could be used instead. Nice thing is that 1. it is just one line instead of 3 lines for the --filesystem finish-args and 2. that it creates the `.org.armagetronad.ArmagetronAdvanced` inside the app sandboxed directory structure instead of directly in the user home directory. With `--persist`, it will be created as `~/.var/app/org.armagetronad.ArmagetronAdvanced/.org.armagetronad.ArmagetronAdvanced` which looks little bad, but it will work correctly and will be safely sandboxed. :-)

zmanuel replied:
That would work and would indeed be easier, but there are two problems you can't know about: The killer for me is that the game is available on different platforms and nobody can remember where configuration and data files are supposed to go on each, so we implemented an ingame help page that shows those directories. With --persist, that page is a lie. Well, it's true within the sandbox, but not outside where the user would typically operate.

Less importantly, we moved the configuration files around for 0.4 already (into ~/.config and ~/.local), but would like to be able to import 0.2.9's config, so then we'd need two --persist lines, one just for compatibility. Or some really good forward thinking now for 0.2.9.

I first tried using a wrapper script to use our command line arguments to bend the directories we use, but now I remembered we also have preprocessor defines one can use to hardcode directories. With environment variable substitution to boot! Not my code :) The only difficulty is getting the ${XDG...} construct unevaluated into the code, past JSON and the build scripts.

AsciiWolf replied:
I have suggested a possible solution in: https://github.com/flathub/org.armagetronad.ArmagetronAdvanced/pull/5/commits/43e63fccfa4f29875f372f6106aeac22b86ad998#r548146855 :-)

AsciiWolf replied:
I have suggested a possible solution in: https://github.com/flathub/org.armagetronad.ArmagetronAdvanced/pull/5/files#r548150401 :-)

AsciiWolf commented on org.armagetronad.ArmagetronAdvanced.json:
```diff
@@ -50,6 +54,16 @@
                 }
+            ],
+            "post-install": [
+                "sed -i /app/share/metainfo/*.appdata.xml -e 's,org.armagetronad.org.armagetronad,org.armagetronad,' -e 's,.desktop</id>,</id>,' -e 's,<!-- <launchable,<launchable,' -e 's,</launchable> -->,</launchable>,'",
```
Protip: The releases section should also be possible to be added/updated by sed, but I am too dumb (and lazy) to do it properly. :-)

zmanuel replied:
I... tend to just pick out a unique substring of the line I want to replace, then replace the whole line with that on with the line I want. It's my hammer. Or rather, that's what I should do, because if I don't, I create mess like org.armagetronad.org.armagetronad.

zmanuel committed:
Store user configuration and data in ${XDG_*_HOME}

instead of using ~/.org.armagetronad.ArmagetronAdvanced, where
we would usually place user specific data with the given
configuration.

Use preprocessor defines to bend user directories.

AsciiWolf commented on org.armagetronad.ArmagetronAdvanced.json:
```diff
@@ -31,7 +28,8 @@
                 "--disable-uninstall",
-                "progname=org.armagetronad.ArmagetronAdvanced"
+                "progname=org.armagetronad.ArmagetronAdvanced",
+                "CXXFLAGS=-DUSER_DATA_DIR='\"$${XDG_DATA_HOME}/game\"' -DUSER_CONFIG_DIR='\"$${XDG_CONFIG_HOME}/game\"' -DAUTORESOURCE_DIR='\"$${XDG_CACHE_HOME}/resource\"'"
```
This is quite nonstandard. Apps usually use their name as a directory inside XDG DATA/CONFIG/CACHE directories.

So, for example, Armagetron could use:

`${XDG_DATA_HOME}/armagetronad`, `${XDG_CONFIG_HOME}/armagetronad` and `${XDG_CACHE_HOME}/armagetronad`

AsciiWolf commented on org.armagetronad.ArmagetronAdvanced.json:
```diff
@@ -22,32 +19,31 @@
+                "--disable-games",
+                "--disable-uninstall",
+                "progname=org.armagetronad.ArmagetronAdvanced",
+                "CXXFLAGS=-DUSER_DATA_DIR='\"$${XDG_DATA_HOME}/game\"' -DUSER_CONFIG_DIR='\"$${XDG_CONFIG_HOME}/game\"' -DAUTORESOURCE_DIR='\"$${XDG_CACHE_HOME}/resource\"'"
```
This is quite nonstandard. Apps usually use their name as a directory inside XDG DATA/CONFIG/CACHE directories.

So, for example, Armagetron could use:

`${XDG_DATA_HOME}/armagetronad`, `${XDG_CONFIG_HOME}/armagetronad` and `${XDG_CACHE_HOME}/armagetronad`

zmanuel replied:
Ach, of course. I hadn't realized those were the environment variables one is meant to use instead of ~/.share, ~/.cache and ~/.config when not in a sandbox. We.... already use that on 0.4, duh. The only choice here then is to make it forward compatible so on 0.4, the special defines can just go away.

zmanuel committed:
Set progname to ArmagetronAdvanced

not the full application id org.armagetronad.ArmagetronAdvanced.
The output of that was not sufficiently correct anyway, and the new
output is more pleasing (shorter, sometimes user visible, path names).
Changes have already been implemented upstream so this variant
can work with less mending for 0.2.9.2.

zmanuel committed:
Adapt user data directories to what 0.4 will do on its own

zmanuel committed:
Use cleanup section for cleanup

zmanuel commented:
I think that should do it.

For 0.2.9.2, I pulled the two patches into the main repository so they're no longer required and fixed the wrong names and content for the installed icons and desktop file; that will make the manifest much simpler: https://github.com/ArmagetronAd/org.armagetronad.ArmagetronAdvanced/blob/0.2.9_alpha/org.armagetronad.ArmagetronAdvanced.json (ignore that it's pulling from a local git repository, obviously). Some hackery of the appinfo.xml will stay required, it also needs to work for AppImage builds on Ubuntu 16.04 because that's how they roll, and they have different demands.

zmanuel reviewed (commented):
(on a removed commit that would clean up /lib/debug)
No, doesn't seem to be required, contents don't land in the regular pack anyway. Used for the optional debug package?
